### PR TITLE
ENH: loadarff now supports relational attributes.

### PR DESCRIPTION
**Pull request by vnmabus**
This solves #9518.

The module arffread has been refactored to a more object-oriented style, so that adding the relational attributes was easier. As a result, some parts where a tuple (name, additional_string) was expected were changed to expect an object.

Relational attributes are variable length lists of instances, each consisting in several attributes. As such, they have `np.object_` dtype.

There are a few controversial points:
- The data is parsed with a csv reader, to allow the use of `'\t'` and `','` as separators, and single or double quotes. The dialect is sniffed from the first line.
- The global csv.field_size_limit is changed to the maximum value. This is because some relational fields can be very long (I have datasets in which that happens). I have not found a way to not using a global variable.
- If the date format is incorrect, a `ParseArffError` will be raised, instead of a `ValueError`. This is because now this error is detected while reading the header.
- ~~The `MetaData` object `__iter__` and `__getitem__` now return attribute objects, instead of names or tuples.~~

This is my first PR, so please report anything that I have done incorrectly.

### CODE REVIEW

**Comment by rgommers** [comment]
> If the date format is incorrect, a `ParseArffError` will be raised, instead of a `ValueError`. This is because now this error is detected while reading the header.

`ParseArffError` inherits from `IOError` instead of `ValueError`, so technically this is a backwards compat break. `IOError` makes sense though, so I think I'm fine with that.

> The data is parsed with a csv reader, to allow the use of `'\t'` and `','` as separators, and single or double quotes. The dialect is sniffed from the first line.

Just to make sure: those are valid ARFF separators, and we weren't supporting them before? No backwards compat break?

> The `MetaData` object `__iter__` and `__getitem__` now return attribute objects, instead of names or tuples.

This looks like a larger breaking change; `MetaData` is public and even the behavior of the example in the docstring is changed (http://scipy.github.io/devdocs/generated/scipy.io.arff.MetaData.html#scipy.io.arff.MetaData)

File: scipy/io/arff/arffread.py [hunk 1]
```diff
@@ -58,168 +64,375 @@ class ArffError(IOError):
...
+
+    raise ParseArffError("unknown attribute %s" % attr_string)
+
+
+def split_data_line(line, dialect=None):
+    delimiters = ",\t"
+
+    # This can not be done in a per reader basis, and relational fields
+    # can be HUGE
+    csv.field_size_limit(int(ctypes.c_ulong(-1).value // 2))
```
Comment (rgommers) [inline scipy/io/arff/arffread.py]: This looks inefficient and changes global state. If you need to play with `field_size_limit`, then it looks to me like you should set it outside the for-loops that call `split_data_line`, and then reset it to the old value when you're done.

**Comment by vnmabus** [comment]
> Just to make sure: those are valid ARFF separators, and we weren't supporting them before? No backwards compat break?

Yes, they are valid (https://waikato.github.io/weka-wiki/arff_stable/#the-instance-data). Previously we were supporting only the comma, and splitting the lines at it. However, the relation attributes have embedded commas, so that this naive approach will not work anymore. That is why I chose to use the csv module to do the work. Since it can support several separators, I have set it up to allow for tab separated values too. Nevertheless, as only one separator will be used, the file must use consistently one separator for this to work, which I think is a reasonable constraint.

> This looks like a larger breaking change; `MetaData` is public and even the behavior of the example in the docstring is changed (http://scipy.github.io/devdocs/generated/scipy.io.arff.MetaData.html#scipy.io.arff.MetaData)

Yes, that is why I was asking if a change of behaviour is correct or no. I will emulate the old behaviour if that is preferable.

File: scipy/io/arff/arffread.py (hunk 1 above)
Comment (vnmabus) [reply scipy/io/arff/arffread.py]: I have though about that, and the problem would be that if two threads do that, then the limit would be reset to the old value before one of them finishes, which is undesirable. As this variable is typically only modified with the objective of increasing it, I think that leaving it set to the maximum value makes sense. Of course, the preferred solution would be a per-reader or per-thread way of doing the same thing, but no one is available in the `csv` module.

**Comment by rgommers** [comment]
> Nevertheless, as only one separator will be used, the file must use consistently one separator for this to work, which I think is a reasonable constraint.

agreed

> Yes, that is why I was asking if a change of behaviour is correct or no. I will emulate the old behaviour if that is preferable.

Yes, that looks preferable - we try not to break existing code unless absolutely necessary.

**Comment by rgommers** [comment]
The CI failure is real - our code style checker can be nitpicky:
```
38.31s$ pycodestyle scipy benchmarks/benchmarks
scipy/io/arff/tests/test_arffread.py:330:37: W292 no newline at end of file
1       W292 no newline at end of file
The command "pycodestyle scipy benchmarks/benchmarks" exited with 1.
```

File: scipy/io/arff/arffread.py (hunk 1 above)
Comment (rgommers) [reply scipy/io/arff/arffread.py]: Okay fair enough. not ideal, but I think I'm fine with changing global state in this case.

**Comment by vnmabus** [comment]
Changes requested done.

**Comment by rgommers** [comment]
Looking through other `loadarff` issues, gh-7446 seems related. This PR doesn't solve that problem though. Unclear to me from a quick read if it should - could you have a look at that @vnmabus?

**Comment by rgommers** [comment]
Also there is gh-3535 for sparse arff support, which stalled a long time ago. May be of interest @vnmabus.

**Comment by vnmabus** [comment]
> Looking through other `loadarff` issues, [gh-7446](https://github.com/scipy/scipy/issues/7446) seems related. This PR doesn't solve that problem though. Unclear to me from a quick read if it should - could you have a look at that @vnmabus?

No, I did not change how dates are parsed. However, I have rewritten the module in a more OOP style, so probably it would be better to address that issue after this PR is merged.

**Comment by vnmabus** [comment]
Reading the comments in the sparse issue, I see that `read_header` is considered public, in spite of it not being in the ´__all__´ list. If that is true, I will have to change it to keep backward compatibility.

**Comment by rgommers** [comment]
And searching my archives for ARFF related stuff, I found https://github.com/renatopp/liac-arff which I think is more fully functional. (this is more a note to self, to have ARFF related things in one place now that I'm looking at it).

**Comment by vnmabus** [comment]
> And searching my archives for ARFF related stuff, I found https://github.com/renatopp/liac-arff which I think is more fully functional. (this is more a note to self, to have ARFF related things in one place now that I'm looking at it).

Ìt does not seem to support relational attributes, though.

**Comment by rgommers** [comment]
> Reading the comments in the sparse issue, I see that read_header is considered public, in spite of it not being in the __all__ list. If that is true, I will have to change it to keep backward compatibility.

No I don't agree with that comment. It looks like a utility function, is explicitly not in `__all__` and not in our docs, so not public.

**Comment by vnmabus** [comment]
Is there anything remaining for me to do to have this pull request accepted?

**Comment by rgommers** [comment]
@vnmabus probably not, I just need to find time for a last round of review. I've marked this for the next release (planned for May) to be sure not to forget.

**Comment by rgommers** [comment]
Okay went through this again, and paid particular attention to backwards compat. There are no changes other than the very minor ones already discussed above.

New test cases and changes look good, so I'll merge. 

Thanks a lot @vnmabus

**Pull request merged**